### ISSUE TITLE
MD013 reflow deletes the space before French double punctuation (`:` `;` `!` `?`) and destroys non-breaking spaces

### ISSUE THREAD

**Issue by oussfyi** [rvben/rumdl#740]
### rumdl version

0.2.34

### Installation method

mise

### Bug description

> [!NOTE]
> This is just a good faith disclaimer that this bug report was authored by Claude Code
> while I was working on my bilingual blog. I had Claude investigate the issue when it came up
> and I did my best to verify its claims, but in the event there's something wrong with this report,
> it's on me and I sincerely apologize for the mistake. Thank you!

When `reflow` re-wraps a paragraph, it strips the space that precedes the
punctuation marks `:` `;` `!` `?`, attaching the mark to the preceding word
(English convention). French typography requires a space before each of these
marks (ideally a narrow no-break space U+202F, or U+00A0). Worse, reflow also
converts every non-breaking space it touches (U+00A0 and U+202F) into an
ordinary breakable space — or, before `:;!?`, deletes it outright — so the
non-breaking guarantee is lost even where the visible spacing survives (e.g. a
`10 000` thousands separator, or the space around guillemets). No lint rule
fires on `" :"`, so this is purely the reflow/normalization pass, and it happens
silently on `rumdl fmt`.

### Steps to reproduce

Input `fr.md` (one line, 143 chars, exceeds the 80-col limit):

```markdown
Voici une phrase en français assez longue pour forcer un retour à la ligne : la ponctuation double doit garder son espace avant le deux-points.
```

Then run the following command (or use the `rumdl.toml` configuration included below) :

```console
$ rumdl fmt --diff --no-config \
    -c 'MD013.line-length = 80' \
    -c 'MD013.reflow = true' \
    -c 'MD013.code-blocks = false' \
    fr.md
```

### Expected behavior

The space (a required French orthographic space) is preserved:

```text
Voici une phrase en français assez longue pour forcer un retour à la ligne : la
ponctuation double doit garder son espace avant le deux-points.
```

### Actual behavior

The space before `:` is deleted (`ligne :` → `ligne:`):

```diff
--- fr.md
+++ fr.md (fixed)
@@ -1,2 +1,2 @@
-Voici une phrase en français assez longue pour forcer un retour à la ligne : la ponctuation double doit garder son espace avant le deux-points.
+Voici une phrase en français assez longue pour forcer un retour à la ligne: la
+ponctuation double doit garder son espace avant le deux-points.
```

### Configuration

```toml
[MD013]
line-length = 80
reflow = true
code-blocks = false
```

### Operating system

Linux

#### Root-cause observations

- **It is the reflow join, not a lint rule.** With `line-length = 500` (no
  rewrap) the space is preserved and `rumdl check` reports nothing on `" :"`;
  with `line-length = 80` (rewrap) the space is deleted. Nothing in `check`
  flags `" :"`, so the deletion happens only in the reflow rewrite path.
- **Position-independent.** The deletion is not a wrap-boundary artifact. A
  colon early in the line (`Court début : puis …`, far from column 80) still
  becomes `Court début: puis …`. So the reflow pass normalizes spacing before
  `:;!?` across the whole paragraph, not just where the line happens to break.
- **Affected marks:** `:` `;` `!` `?` — for each, the preceding space is
  removed entirely (the mark is glued to the previous word).
- **Not affected:** guillemets — `«` (space after) and `»` (space before) keep
  their surrounding ASCII space.
- **Non-breaking spaces are destroyed regardless of the mark.** Putting U+00A0
  or U+202F before the colon yields the same `ligne:` (space gone). A lone
  U+00A0 elsewhere — a `10 000` thousands separator, or the space before `»` —
  survives *visually* but is downgraded to an ordinary breakable space (the
  U+00A0 byte count drops to 0 after `fmt`). So reflow re-tokenizes on
  whitespace and rejoins with plain ASCII spaces, discarding non-breaking
  spaces.
- **Distinct from ordinary space collapsing.** A double space between two plain
  words collapses to a single space (expected reflow behavior). The `:;!?` case
  is different: the space is removed, not collapsed to one.
- **`reflow-mode` interaction.** `default` and `normalize` both strip the space
  (and NBSP). `sentence-per-line` and `semantic-line-breaks` preserve the space
  and the NBSP — but they abandon column-fill (they leave the over-length line
  unwrapped / break only at sentence or clause boundaries), so they are not a
  workaround for someone who wants ordinary 80-column wrapping.

#### Impact

Silent corruption of French (and any locale using spaced double punctuation)
prose on `rumdl fmt`: every `mot :` / `mot ;` / `mot !` / `mot ?` becomes
typographically wrong `mot:` / `mot;` / `mot!` / `mot?`, and every non-breaking
space in reflowed text loses its non-breaking property (allowing bad line
breaks such as `10` / `000`). No warning is emitted; the damage is only visible
by diffing content. This blocks using MD013 reflow on bilingual/French
Markdown.

#### Duplicate check

Appears **novel**. Searched the tracker (open + closed) via `gh issue list
--search` on `reflow`, `reflow space`, `reflow whitespace`, `french`,
`punctuation`, `nbsp`, `non-breaking space`, `insécable`, `space before colon`,
`colon reflow`, `MD064`, `sentence spacing`; also WebSearch and the CHANGELOG.
Many reflow bugs exist but none concern space-before-`:;!?` deletion, French
double punctuation, or non-breaking-space handling. Related but distinct: #597
(reflow *inserts* a space before literal `*` — opposite direction), #160
(sentence-per-line English abbreviation exceptions), #202 (feature request to
collapse multiple spaces), and the MD064 sentence-spacing rule family. No fix
after 0.2.34 (it is the latest release).

**Comment by rvben** [comment]
Thanks for the thorough report, and no need to apologize - it reproduced exactly.

Both parts are fixed in v0.2.35 (released after 0.2.34):

- The space before French double punctuation (`:` `;` `!` `?`) is preserved through reflow.
- Non-breaking spaces (U+00A0 and U+202F) survive reflow instead of being downgraded to ordinary spaces, so a `10 000` thousands separator stays non-breaking and no longer breaks across lines.

Fixed in f66021f0. Verified on 0.2.35: your `fr.md` wraps to `... ligne : la` with the space intact, and a U+00A0 separator keeps its bytes after `fmt`.

Upgrade with `mise` (or `cargo install rumdl`) to 0.2.35. Thanks again.
